external-crates/move/crates/move-binary-format/src/compatibility.rs

```diff
@@ -360,98 +362,165 @@ pub enum InclusionCheck {
-            && (old_module.structs.len() != new_module.structs.len()
-                || old_module.enums.len() != new_module.enums.len()
-                || old_module.functions.len() != new_module.functions.len()
-                || old_module.friends.len() != new_module.friends.len())
-        {
-            return err;
+        if old_module.address != new_module.address {
+            context.module_address_mismatch(&old_module.address, &new_module.address);
         }
 
-        // Struct checks
-        for (name, old_struct) in &old_module.structs {
-            match new_module.structs.get(name) {
-                Some(new_struct) if old_struct == new_struct => (),
-                _ => {
-                    return err;
-                }
-            };
+        if old_module.file_format_version > new_module.file_format_version {
+            context.file_format_version_downgrade(
+                old_module.file_format_version,
+                new_module.file_format_version,
+            );
         }
 
-        // Enum checks
-        for (name, old_enum) in &old_module.enums {
-            let Some(new_enum) = new_module.enums.get(name) else {
-                return err;
-            };
-
-            if old_enum.abilities != new_enum.abilities {
-                return err;
-            }
-            if old_enum.type_parameters != new_enum.type_parameters {
-                return err;
-            }
-            if old_enum.variants.len() > new_enum.variants.len() {
-                return err;
+        // Since the structs are sorted we can iterate through each list to find the differences using two pointers
+        for mark in compare_identifiers(old_module.structs.keys(), new_module.structs.keys()) {
+            match mark {
+                Mark::New(name) => context.struct_new(name, new_module.structs.get(name).unwrap()),
+                Mark::Missing(name) => context.struct_missing(name),
+                Mark::Existing(name) => {
+                    let old_struct = old_module.structs.get(name).unwrap();
+                    let new_struct = new_module.structs.get(name).unwrap();
+                    if old_struct != new_struct {
+                        context.struct_change(name, old_struct, new_struct);
+                    }
+                }
             }
-
-            // NB: In the future if we allow adding new variants to enums in subset mode
-            // remove this if statement. This check is somewhat redundant with the one
-            // below, the one below should be kept if we allow adding variants in subset
-            // mode.
-            if old_enum.variants.len() != new_enum.variants.len() {
-                return err;
+        }
+        // enum checks
+        compare_identifiers(old_module.enums.keys(), new_module.enums.keys()).for_each(|mark| {
+            match mark {
+                Mark::New(name) => context.enum_new(name, new_module.enums.get(name).unwrap()),
```

Comment:
Why not iterate over the key value pairs and return references to the values in your `Mark` struct to avoid the `unwrap`s?